src/dev/code_coverage/ingest_coverage/ingest.js

```diff
@@ -19,40 +19,74 @@
+import { errMsg, redact, whichIndex } from './ingest_helpers';
+import { pretty, green } from './utils';
 import { right, left } from './either';
 
 const node = process.env.ES_HOST || 'http://localhost:9200';
+
 const client = new Client({ node });
-const pipeline = process.env.PIPELINE_NAME || 'team_assignment';
-const redacted = redact(node);
+const redactedEsHostUrl = redact(node);
+const parse = JSON.parse.bind(null);
+const isResearchJob = process.env.COVERAGE_JOB_NAME === RESEARCH_CI_JOB_NAME ? true : false;
 
 export const ingest = (log) => async (body) => {
-  const index = body.isTotal ? TOTALS_INDEX : COVERAGE_INDEX;
-  const maybeWithPipeline = maybeTeamAssign(index, body);
-  const withIndex = { index, body: maybeWithPipeline };
-  const dontSend = noop;
-
-  log.verbose(withIndex);
-
-  process.env.NODE_ENV === 'integration_test'
-    ? left(null)
-    : right(withIndex).fold(dontSend, async function doSend(finalPayload) {
-        await send(index, redacted, finalPayload);
-      });
+  const isTotal = !!body.isTotal;
+  const index = whichIndex(isResearchJob)(isTotal);
+  const isACoverageIndex = isTotal ? false : true;
+
+  const stringified = pretty(body);
+  const pipeline = TEAM_ASSIGNMENT_PIPELINE_NAME;
+
+  const finalPayload = isACoverageIndex
+    ? { index, body: stringified, pipeline }
+    : { index, body: stringified };
+
+  const justLog = dontSendButLog(log);
+  const doSendToIndex = doSend(index);
+  const doSendRedacted = doSendToIndex(redactedEsHostUrl)(log)(client);
+
+  eitherSendOrNot(finalPayload).fold(justLog, doSendRedacted);
 };
 
-async function send(idx, redacted, requestBody) {
+function doSend(index) {
+  return (redactedEsHostUrl) => (log) => (client) => async (payload) => {
+    const logF = logSend(true)(redactedEsHostUrl)(log);
+    await send(logF, index, redactedEsHostUrl, client, payload);
+  };
+}
+
+function dontSendButLog(log) {
+  return (payload) => {
+    logSend(false)(null)(log)(payload);
+  };
+}
+
+async function send(logF, idx, redactedEsHostUrl, client, requestBody) {
   try {
     await client.index(requestBody);
+    logF(requestBody);
   } catch (e) {
-    throw createFailError(errMsg(idx, redacted, requestBody, e));
+    const { body } = requestBody;
+    const parsed = parse(body);
+    throw createFailError(errMsg(idx, redactedEsHostUrl, parsed, e));
   }
 }
 
-export function maybeTeamAssign(index, body) {
-  const payload = index === TOTALS_INDEX ? body : { ...body, pipeline };
-  return payload;
+function logSend(actuallySent) {
+  return (redactedEsHostUrl) => (log) => (payload) => {
+    const { index, body } = payload;
+    log.verbose(
+      `### ${actuallySent ? 'Sent' : 'Fake Sent'}:
+${redactedEsHostUrl ? `\t### ES Host: ${redactedEsHostUrl}` : ''}
+\t### Index: ${green(index)}
+\t### payload.body: ${body}
+${process.env.NODE_ENV === 'integration_test' ? `ingest-pipe=>${payload.pipeline}` : ''}
+`
+    );
+  };
+}
```

Comment:
imho this function not the easy to read and its call looks "heavy" `logSend(false)(null)(log)(payload);`
since it still around the same thing, how about just pass an object/array of args ?

Reply:
Well, I'm using partial application here to pass around "large" objects like the logger.  So, putting all the args in one function w/o using function currying would make have to figure out of a way pass those around differently.  Not sure if it's worth the time.  I could break the message building out of the function, to make it smaller?  Whatcha think?  We can zoom about it too

Reply:
@dmlemeshko https://javascript.info/currying-partials <- explains it better than I can lol